File: DP/1449. Form Largest Integer With Digits That Add up to Target/1449. Form Largest Integer With Digits That Add up to Target.java
class Solution {
    public String largestNumber(int[] cost, int target) {
        int n = cost.length;
        String[] dp = new String[target+1];
        
        dp[0] = "";
        for (int c = 1; c <= target; c++) {
            dp[c] = "#";
            for (int i = 1; i <= n; i++) {
                if (c < cost[i-1] || dp[c-cost[i-1]].equals("#")) continue;
                String cur = dp[c-cost[i-1]] + (char)(i+'0');
                
                if (cur.length() > dp[c].length()) {
                    dp[c] = cur;
                } else if (cur.length() == dp[c].length()) {
                    if (dp[c].compareTo(cur) < 0) {
                        dp[c] = cur;
                    }
                }   
            }
        }

        if (dp[target].equals("#")) return "0";
        else return dp[target];
    }
}

// dp[i][j]: the maximum integer you can paint with the ith object and j capacity

// unlimited knapsnack
// dp[i][j] = max(dp[i-1][j], dp[i][j-cost[i]]+w[i])
// String[][] dp = new String[n+1][target+1];
//     for (int c = 1; c <= target; c++) {
//         for (int i = 1; i <= n; i++) {
//             dp[i][c] = dp[i-1][c];
//             if (c-cost[i-1] >= 0) {
//                 String cur = dp[i][c-cost[i-1]] + 'i';
//                 if (dp[i][c].compareTo(cur) < 0) {
//                     dp[i][c] = cur;
//                 }
//             }
            
//         }
//     }
// return dp[n][target];

File: DP/1449. Form Largest Integer With Digits That Add up to Target/1449. Form Largest Integer With Digits That Add up to Target.cpp
class Solution {
public:
    string largestNumber(vector<int>& cost, int target) {
        string dp[target+1];
        for (int c = 1; c <= target; c++) {
            dp[c] = "#";
            for (int i = 1; i <= 9; i++) {
                if (c < cost[i-1] || dp[c-cost[i-1]] == "#") continue;
                string cur = dp[c-cost[i-1]] + (char)(i+'0');
                
                if (cur.size() > dp[c].size() || (cur.size() == dp[c].size() && cur > dp[c])) {
                    dp[c] = cur;
                } 
            }
        }

        if (dp[target] == "#") return "0";
        else return dp[target];
    }
};

// 01 Knapscak

// obj[i] cap
// dp[c]: the maximun goal if using c
// for i in obj
//     for c in capacity
//         dp[i][c] = max(dp[i-1][c], dp[i-1][c-cost[i]]+vali[i]);
// return max{dp[n][c]} for c = ....
    
// unlimited knapsack
// for c in capacity
//     for i in obj
//         dp[c] = max(dp[c], dp[c-cost[i]]+vali[i]);
// return max{dp[c]} for c = ....
